src/machine/asm/mod.rs

```diff
@@ -303,14 +303,17 @@ impl SupportMachine for Box<AsmCoreMachine> {
         Some(self.max_cycles)
     }
 
-    fn reset(&mut self) {
+    fn reset(&mut self, max_cycles: Option<u64>) {
         self.registers = [0; RISCV_GENERAL_REGISTER_NUMBER];
         self.pc = 0;
         self.flags = [0; RISCV_PAGES];
         for i in 0..TRACE_SIZE {
             self.traces[i] = Trace::default();
         }
         self.frames = [0; MEMORY_FRAMES];
+        self.cycles = 0;
+        debug_assert!(max_cycles.is_some());
+        self.max_cycles = max_cycles.unwrap_or(0);
         self.reset_signal = 1;
```

Comment:
If I read correctly when reset_signal is 1, the aot code will be discarded, so the behavior of aot machine should be problematic

https://github.com/nervosnetwork/ckb-vm/blob/develop/src/machine/asm/mod.rs#L371-L373

Reply:
Yes, in this case, aot will be converted to asm to run. This is because it is not possible to add new aot code for SupportMachine in the exec system call .

Reply:
Ignore exec syscall first.

If it is a normal aot VM, the behavior of losing aot code after reset seems a bit weird. User expects reset to return to the initialization state with aot code?

Reply:
> This is because it is not possible to add new aot code for SupportMachine in the exec system call .

We should fix this instead.

Reply:
In this case, you can't just `reset()` and then `load_program()`, because the aot code is not generated from `load_program`. The correct approach is:

```
machine.reset()
let mut aot_machine = AotCompilingMachine::load(_);
let code = aot_machine.compile().unwrap();
machine.aot_code = code
machine.load_program(_);
machine.run();
```

Reply:
> We should fix this instead. Now that we think about it, it would make more sense to provide AoT code, when we are doing load_program, not when we are initializing the VM.

We can solve this problem in the next PR

Reply:
A different PR works, but we should definitely fix this.